huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go

```diff
@@ -0,0 +1,450 @@
+				Computed: true,
+			},
+
+			"auth_type": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"node_type": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func ResourceDeviceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	createOpts := buildDeviceCreateParams(d)
+	resp, err := client.AddDevice(createOpts)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA device: %s", err)
+	}
+
+	if resp.DeviceId == nil {
+		return diag.Errorf("error creating IoTDA device: id is not found in API response")
+	}
+
+	d.SetId(*resp.DeviceId)
+
+	// bind tags
+	err = bindDeviceTags(client, d.Id(), nil, d.Get("tags").(map[string]interface{}))
+	if err != nil {
+		return diag.Errorf("error binding tags when creating IoTDA device: %s", err)
+	}
+
+	// freeze device
+	isFrozen := d.Get("frozen").(bool)
+	err = updateDeviceStatus(client, d.Id(), utils.StringValue(resp.Status), isFrozen)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return ResourceDeviceRead(ctx, d, meta)
+}
+
+func ResourceDeviceRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	response, err := client.ShowDevice(&model.ShowDeviceRequest{DeviceId: d.Id()})
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving IoTDA device")
+	}
+
+	mErr := multierror.Append(
+		d.Set("region", region),
+		d.Set("name", response.DeviceName),
+		d.Set("device_id", response.DeviceId),
+		d.Set("node_id", response.NodeId),
+		d.Set("name", response.DeviceName),
+		d.Set("product_id", response.ProductId),
+		d.Set("gateway_id", response.GatewayId),
+		d.Set("description", response.Description),
+		d.Set("space", response.AppId),
+		d.Set("status", response.Status),
+		d.Set("node_type", response.NodeType),
+		d.Set("tags", flattenTags(response.Tags)),
+		d.Set("frozen", utils.StringValue(response.Status) == deviceStatusFrozen),
+	)
+
+	if response.AuthInfo != nil {
+		mErr = multierror.Append(mErr,
+			d.Set("secret", response.AuthInfo.Secret),
+			d.Set("fingerprint", response.AuthInfo.Fingerprint),
+			d.Set("auth_type", response.AuthInfo.AuthType),
+		)
+	}
+
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+func ResourceDeviceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	// name,desc
+	updateOpts := buildDeviceUpdateParams(d)
```

Comment:
should add d.HasChanges("name", "description") check

Reply:
done